### PR TITLE
Add maybe_unused to ranges and nodiscard to start_range.

### PR DESCRIPTION
**Pull request by jrhemstad**
Some compilers issues spurious warnings about unused variables for
instances of scoped_range and unique_range. Use [[maybe_unused]] when available
to silence these warnings.

Likewise, added `[[nodiscard]]` when available to `start_range()` to issue a warning
if a user accidently forgets to store the result of `start_range()`. It is not possible to correctly
use `start_range()` without using the result to call `end_range()`.

### CODE REVIEW

**Comment by jrhemstad** [comment]
Hey @jcohen-nvidia would you mind reviewing this so we can get it merged?

File: c/include/nvtx3/nvtx3.hpp [hunk 1]
```diff
@@ -902,7 +915,7 @@ class domain {
    * "global" NVTX domain.
    *
    */
-  domain() noexcept {}
+  constexpr domain() noexcept {}
```
Comment (evanramos-nvidia) [inline c/include/nvtx3/nvtx3.hpp:920]: This `constexpr` depends on C++11, and will also generate warnings on older versions of MSVC.

File: c/include/nvtx3/nvtx3.hpp [hunk 2]
```diff
@@ -611,6 +611,19 @@
...
 
+// Macro wrappers for C++ attributes: [[maybe_unused]] and [[nodiscard]]
+#if defined(__has_cpp_attribute)
+  #if __has_cpp_attribute(maybe_unused)
+    #define NVTX3_MAYBE_UNUSED [[maybe_unused]]
+  #endif
+  #if __has_cpp_attribute(nodiscard)
+    #define NVTX3_NO_DISCARD [[nodiscard]]
+  #endif
+#else
```
Comment (evanramos-nvidia) [inline c/include/nvtx3/nvtx3.hpp]: These need preprocessor checks for the C++ language version that implements these attributes, and we could also benefit from adding compiler-specific alternatives for broader support.

**Review by evanramos-nvidia — CHANGES_REQUESTED** [review]

File: c/include/nvtx3/nvtx3.hpp (hunk 1 above)
Comment (evanramos-nvidia) [reply c/include/nvtx3/nvtx3.hpp:920]: Nevermind this point, I see we already use `constexpr` constructors throughout the file.

**Pull request merged**